File: ESTRUCTURA DE DATOS/DEBERES/SEGUNDO_PARCIAL/DEBER2/PROYECTO_NRC1978/LISTA_SIMPLE/Propietario.cpp
#include "Propietario.h"
#include "Validaciones.h"
#include "ListaPropietarios.h"
#include <algorithm>
#include "ValidacionPlaca.h"


using namespace std;

Propietario::Propietario()
    : nombre(""), apellido(""), cedula(""), correo("") {}


Propietario::Propietario(string nombre, string apellido, string cedula, string correo)
    : nombre(nombre), apellido(apellido), cedula(cedula), correo(correo) {}


void Propietario::setNombre(const string &nombre)
{
    this->nombre = nombre;
}

void Propietario::setApellido(const string &apellido)
{
    this->apellido = apellido;
}

void Propietario::setCedula(const string &cedula)
{
    this->cedula = cedula;
}

void Propietario::setCorreo(const string &correo)
{
    this->correo = correo;
}

void Propietario::eliminarPlaca(const string &placa)
{
    auto it = find(placas.begin(), placas.end(), placa);
    if (it != placas.end())
    {
        placas.erase(it);
        cout << "Placa eliminada exitosamente." << endl;
    }
    else
    {
        cout << "Placa no encontrada." << endl;
    }
}

string Propietario::getNombre() const
{
    return nombre;
}

string Propietario::getApellido() const
{
    return apellido;
}

string Propietario::getCedula() const
{
    return cedula;
}

string Propietario::getCorreo() const
{
    return correo;
}



void Propietario::agregarPlaca(const string &placa)
{
    placas.push_back(placa);
}

vector<string> Propietario::getPlacas() const
{
    return placas;
}


string Propietario::toString() const
{
    string result = "Nombre: " + nombre + ", Apellido: " + apellido + ", Cedula: " + cedula + ", Correo: " + correo +  ", Placas: ";
    for (const auto &placa : placas)
    {
        result += placa + " ";
    }
    return result;
}


void agregarPropietario(ListaPropietarios &listaPropietarios)
{
    Validaciones validaciones;
    string nombre = validaciones.ingresarString("Ingrese el nombre del propietario: ");
    string apellido = validaciones.ingresarString("Ingrese el apellido del propietario: ");
    string cedula;
    do {
        cedula = validaciones.ingresarCedula("Ingrese la cedula del propietario: ");
        if (!validaciones.validarCedula(cedula)) {
            cout << "Cedula invalida. Intente de nuevo." << endl;
        }
    } while (!validaciones.validarCedula(cedula));
    
    string correo;
    do {
        correo = validaciones.ingresarCorreo("Ingrese el correo electronico del propietario: ");
        if (!validaciones.validarCorreo(correo)) {
            cout << "Correo invalido. Intente de nuevo." << endl;
        }
    } while (!validaciones.validarCorreo(correo));


    Propietario nuevoPropietario(nombre, apellido, cedula, correo);
    listaPropietarios.agregarPropietario(nuevoPropietario);

    cout << "Propietario agregado exitosamente." << endl;
}

void eliminarPlacaPropietario(ListaPropietarios &listaPropietarios)
{
    Placa<Coche> validador;
    Validaciones validaciones;
    string cedula = validaciones.ingresarCedula("Ingrese la cedula del propietario: ");
    Propietario* propietario = listaPropietarios.buscarPropietarioPorCedula(cedula);
    if (propietario != nullptr) {
        string placa = validador.ingresarPlaca(nullptr);
        propietario->eliminarPlaca(placa);
    } else {
        cout << "Propietario no encontrado." << endl;
    }
    system("pause");
}

File: ESTRUCTURA DE DATOS/DEBERES/SEGUNDO_PARCIAL/DEBER2/PROYECTO_NRC1978/LISTA_SIMPLE_PYTHON/PYTHON/propietario.py
from validaciones import Validaciones

class Propietario:
    def __init__(self, nombre="", apellido="", cedula="", correo=""):
        self.nombre = nombre
        self.apellido = apellido
        self.cedula = cedula
        self.correo = correo
        self.placas = []

    def set_nombre(self, nombre):
        self.nombre = nombre

    def set_apellido(self, apellido):
        self.apellido = apellido

    def set_cedula(self, cedula):
        self.cedula = cedula

    def set_correo(self, correo):
        self.correo = correo

    def eliminar_placa(self, placa):
        if placa in self.placas:
            self.placas.remove(placa)
            print("Placa eliminada exitosamente.")
        else:
            print("Placa no encontrada.")

    def get_nombre(self):
        return self.nombre

    def get_apellido(self):
        return self.apellido

    def get_cedula(self):
        return self.cedula

    def get_correo(self):
        return self.correo

    def agregar_placa(self, placa):
        self.placas.append(placa)

    def get_placas(self):
        return self.placas

    def to_string(self):
        result = f"Nombre: {self.nombre}, Apellido: {self.apellido}, Cedula: {self.cedula}, Correo: {self.correo}, Placas: "
        result += " ".join(self.placas)
        return result

    def agregar_propietario(lista_propietarios):
        from lista_propietarios import ListaPropietarios
        validaciones = Validaciones()
        nombre = validaciones.ingresar_string("Ingrese el nombre del propietario: ")
        apellido = validaciones.ingresar_string("Ingrese el apellido del propietario: ")
        cedula = ""
        while True:
            cedula = validaciones.ingresar_cedula("Ingrese la cedula del propietario: ")
            if validaciones.validar_cedula(cedula):
                break
            print("Cedula invalida. Intente de nuevo.")
    
        correo = ""
        while True:
            correo = validaciones.ingresar_correo("Ingrese el correo electronico del propietario: ")
            if validaciones.validar_correo(correo):
                break
            print("Correo invalido. Intente de nuevo.")

        nuevo_propietario = Propietario(nombre, apellido, cedula, correo)
        lista_propietarios.agregar_propietario(nuevo_propietario)

        print("Propietario agregado exitosamente.")

    def eliminar_placa_propietario(lista_propietarios):
        from lista_propietarios import ListaPropietarios
        validador = Validaciones()
        validaciones = Validaciones()
        cedula = validaciones.ingresar_cedula("Ingrese la cedula del propietario: ")
        propietario = ListaPropietarios.buscar_propietario_por_cedula(cedula)
        if propietario:
            placa = validador.ingresar_placa(None)
            Propietario.eliminar_placa(placa)
        else:
            print("Propietario no encontrado.")
        input("Presione cualquier tecla para continuar...")

